### ISSUE TITLE
Strategy for tests that use local files

### ISSUE THREAD

**Issue by tintoy** [tintoy/corefx#9]
Relates to problems found in dotnet/corefx#16657.

Maybe a base class for tests that creates a subdirectory named after the test class with methods child tests can call to create files. Those files created in subdirectories of test subdirectory. Auto deleted when disposed (or when test class is disposed). 

Sorry typing on phone so sound like telegram. What do you think, @anthonylangsworth?

**Comment by anthonylangsworth** [comment]
I don't have a problem with the idea. I am just wondering whether we can do better ... and its late and my brain is fried after stepping through the internals of `System.Xml`.

**Comment by peterwurzinger** [comment]
Well to me that sounds like a plan. Maybe concat the name of the directory with a Guid.NewGuid() to ensure not to run into some exceptions after a dirty/incomplete test run.

**Comment by tintoy** [comment]
I'll look into this today or tomorrow. Should be pretty simple.

**Comment by tintoy** [comment]
Something like:

```csharp
class TestWithFilesBase
    : IDisposable
{
    private DirectoryInfo BaseDirectory { get; } = new DirectoryInfo(
        Path.Combine(Directory.GetCurrentDirectory(), "TestFiles", GetType().Name)
    );
    private DirectoryInfo TestDirectory(string testName) => new DirectoryInfo(
        Path.Combine(BaseDirectory.FullName, testName)
    );

    public void Dispose()
    {
        foreach (TempFile tempFile in TempFiles)
            tempFile.Dispose();
    }

    private List<TempFile> TempFiles { get; } = new List<TempFile>();

    protected IDisposable CreateTestFile(string content, string extension, [CallerMemberName] string testName = null)
    {
        return CreateTestFileCore(testName + extension, content);
    }

    IDisposable CreateTestFileCore(string fileName, string content)
    {
        var file = new TempFile(
            Path.Combine(Directory.GetCurrentDirectory(), fileName)
        );

        TempFiles.Add(file);

        File.WriteAllText(file.Path, content);

        return file;
    }
}
```

How does that sound?

**Comment by tintoy** [comment]
@anthonylangsworth @peterwurzinger what do you think of this first attempt (https://github.com/tintoy/corefx/commit/707e9d35c7926c8e79591fe14731cb2a7c5ee43b)?

Directory per test class, and then directory per test method. IIRC, Xunit won't run tests from the same class in parallel, so thread-safety should not be an issue.

**Comment by tintoy** [comment]
Example usage:

```csharp
[Fact]
public void SomeTest()
{
    FileInfo tempFile1 = CreateTempFile("doc.dtd", content: "<!-- presence required, not content -->");
    FileInfo tempFile2 = CreateTempFile("world.txt", content: "world");
    using (EnterTempDirectory())
    {
        // Able to use "doc.dtd" and "world.txt" without qualified path names.
    }
}
```

**Comment by tintoy** [comment]
I guess my only question then is, if running tests in parallel, how is current directory affected? Are they all in the same process? That'd be painful.

**Comment by peterwurzinger** [comment]
Looks good to me! I only could review it on my smartphone though.

**Comment by tintoy** [comment]
If so, then we'll need to get rid of the EnterTempDirectory and modify samples that require temp files to use fully-qualified paths.

For many of the tests, we can use an `XmlResolver` variant such as `XmlPreloadedResolver` which means we can modify them to not need temp files at all, but a couple of them are specifically named to indicate that they're designed to run with no resolver configured (e.g. `C14NSpecExample1_WithoutResolver`).

For the rest, it's as simple as:

```csharp
XmlPreloadedResolver resolver = new XmlPreloadedResolver();
resolver.Add("doc.dtd", "<!-- presence required, rather than content -->");
resolver.Add("world.txt", "world");
```

CC: @bartonjs

**Comment by peterwurzinger** [comment]
Whew, that's a good question. My shot in the dark would be yes, they're running in the same process. Left to prove ofc.

**Comment by tintoy** [comment]
Checked the xUnit docs:

> Tests within the same test class will not run in parallel against each other

So I think we're good :)

**Comment by peterwurzinger** [comment]
Hm. This sounds surprising to me, considering how many problems I already had to face in the past because tests were running parallel. Could have changed ofc.
If this still causes problems, I remember that there is a switch in app.config to disable parallel execution at all.

**Comment by tintoy** [comment]
Of course, there is another much simpler way to deal with all this - if we put all tests that use the `doc.dtd` file in the same test collection (named, for example, "Uses doc.dtd") then xUnit will guarantee that those tests are never run in parallel. Then we can create a collection-level fixture for `doc.dtd`. Ditto for `world.txt`.

@bartonjs - I think this option is a lot simpler but I'm not sure if there's a preferred approach. Any input would be welcome :)

**Comment by tintoy** [comment]
CC: @krwq

**Comment by krwq** [comment]
@tintoy @peterwurzinger @anthonylangsworth could you take a look at my PR and check if this is still valid? I wouldn't bother converting from file system just yet, let's get everything else to work first (unless there are some issues with that solution)

**Comment by tintoy** [comment]
@krwq - I think the issue would still apply; the problem is that 2 different test classes use the same file names. Since xUnit won't run tests from the same class in parallel (but will run tests from different classes in parallel) we really only need to either use different file names per class or use a test collection to prevent parallelisation of test classes that use those files.

I reckon that's a simplest change with the least amount of code. We can always look at using `XmlPreloadedResolver` as a later refactoring step.

**Comment by tintoy** [comment]
(in other words, I'm suggesting we don't use the base class I created in https://github.com/tintoy/corefx/commit/707e9d35c7926c8e79591fe14731cb2a7c5ee43b but instead just ensure those couple of tests don't run in parallel)

**Comment by krwq** [comment]
@tintoy isn't TempFile generating random file names or at least use directory per test or something like that? If not we can make it do so

**Comment by tintoy** [comment]
No - it’s just a wrapper around `try { xxx } finally { File.Delete(tempFileName); }`. I think changing that would impact a lot of other tests as well.

**Comment by krwq** [comment]
@tintoy let's make a private copy then. Doesn't the file name include test name and class though? that should be fairly unique - isn't it?

**Comment by tintoy** [comment]
The implementation I used above uses `TempDirectory` which is like `TempFile` but nukes the entire directory on disposal. It does use class and test name for uniqueness. We'll need to use absolute paths in entity references unless we can change the `XmlResolver` being used.

**Comment by tintoy** [comment]
But the class above will do what you want - see the usage example (except the bit about `EnterTempDirectory()`).

**Comment by tintoy** [comment]
Honestly I'd rather use `XmlPreloadedResolver` wherever possible rather than trying to hack together a solution that can be broken if any test happens to execute `Directory.SetCurrentDirectory` (short of using absolute paths in entity names).

**Comment by krwq** [comment]
@tintoy - look at my PR - i already do that (full path) - that's why the escaping issue showed up.

The more abstraction we add to the test the harder it will be to debug

**Comment by tintoy** [comment]
Sorry, I looked in the PR but can't find any references to a custom resolver. Would you mind attaching a link? Small screen makes it hard to view diffs :-)

**Comment by krwq** [comment]
@tintoy meant full path not custom resolver :]

**Comment by tintoy** [comment]
Ah, sorry. Got it now.

RE: Abstraction - I agree there's a break-even point and beyond that it's all downhill. But I believe that the reason for abstractions in tests is so that it's clear what the test is testing (not necessarily _how_ they're being tested). If it was anywhere else besides corefx I'd be fine with stuff that uses local files, but the build / test process here is kinda byzantine and I'm worried that it'd be hard for people maintaining these tests to reason about why local file-system access doesn't behave as they'd expect.

Base class might be over-engineered, but it'd also be kinda icky to have a bunch of boilerplate in every test just to create a temp file...

**Comment by tintoy** [comment]
Oh well, back to the drawing board perhaps :-)

**Comment by krwq** [comment]
@tintoy - we can tackle on that at any time after we will be feature complete. There are still couple of tests I'm not sure why they are failing

**Comment by tintoy** [comment]
Look at the way regular VSTest tests work. There's a test-run directory, and there's a `TestContext` where you can get access to info about the current test-run's directory.

There's no such thing in xUnit that I can find. So it's just implicit that the current directory is the test directory. Combine with parallelism, and tests that use the file-system are potentially painful...

Cheapest change we can make for files is to just prevent parallelism where it causes problems, then optimise afterwards.

**Comment by tintoy** [comment]
Snap! Yep, no argument here.

**Comment by bartonjs** [comment]
In the X509 tests there's an Xunit collection used for the (okay, only one) class that is depending on filesystem state.

My personal preference is always to eliminate IO from the places where it isn't important to a test.  Just removes a layer of things that can go wrong and create noise.  But if it's a huge amount of work, might not be worth it.

Not much of an answer, I know :smile:.

**Comment by tintoy** [comment]
Defo agree, I think we can probably get rid of 80-90% of the dependency on files; if there's only one that needs it, we can put it in its own test collection and the concurrency issue goes away (hopefully).

**Comment by anthonylangsworth** [comment]
@bartonjs @krwq @tintoy @peterwurzinger I have (hopefully) resolved the concurrency issues with https://github.com/krwq/corefx/pull/1.
